docs/notebooks/pywbemmock.ipynb

```diff
@@ -339,27 +339,65 @@
+    "    parameter. Test for existence of method named method1\n",
     "    \"\"\"\n",
     "\n",
-    "    # Access input parameters\n",
-    "    print('Callback received params: %r' % params )\n",
-    "    print('Callback recieved object_name %s' % objectname)\n",
-    "    ip1 = params['InputParam1']\n",
-    "\n",
-    "    # Access the mock repository through the faked connection object.\n",
-    "    cl = conn.GetClass(objectname)\n",
-    "\n",
-    "    # Set return value == 0 and output parameters\n",
-    "    rtn_val = 0\n",
-    "    # setup output parameter that will be returned to caller\n",
-    "    op1 = CIMParameter('OutputParam11', 'string', value='Some output data')\n",
-    "    return rtn_val, [op1]\n",
+    "    provider_classnames = 'TST_Class'\n",
+    "\n",
+    "    def __init__(self, cimrepository):\n",
+    "        super(Method1TestProvider, self).__init__(cimrepository)\n",
+    "\n",
+    "    def InvokeMethod(self, namespace, MethodName, ObjectName, Params):\n",
+    "        \"\"\"\n",
+    "        Simplistic test method. Validates methodname, objectname, Params\n",
+    "        and returns rtn value 0 and one parameter\n",
+    "\n",
+    "        The parameters and return for Invoke method are defined in\n",
+    "        :meth:`~pywbem_mock.MethodProvider.InvokeMethod`\n",
+    "        \"\"\"\n",
+    "        # validate namespace using method in BaseProvider\n",
+    "        self.validate_namespace(namespace)\n",
+    "\n",
+    "        # get classname and validate. This provider uses only one class\n",
+    "        if isinstance(ObjectName, six.string_types):\n",
+    "            classname = ObjectName\n",
+    "        else:\n",
+    "            classname = ObjectName.classname\n",
+    "        assert classname.lower() == 'method1'\n",
+    "\n",
+    "        # Test if class exists.\n",
+    "        if not self.class_exists(namespace, classname):\n",
+    "            raise CIMError(\n",
+    "                CIM_ERR_NOT_FOUND,\n",
+    "                _format(\"class {0|A} does not exist in CIM repository, \"\n",
+    "                        \"namespace {1!A}\", classname, namespace))\n",
+    "\n",
+    "        if isinstance(ObjectName, CIMInstanceName):\n",
+    "            instance_store = self.cimrepository.get_instance_store(namespace)\n",
+    "            if not instance_store.get(ObjectName):\n",
+    "                raise CIMError(\n",
+    "                    CIM_ERR_NOT_FOUND,\n",
+    "                    _format(\"Instance {0|A} does not exist in CIM repository\",\n",
+    "                            ObjectName))\n",
+    "\n",
+    "        # Execute provider specific code and define any return parameters and\n",
+    "        # a return value\n",
+    "        return_params = None\n",
+    "        return_value = None\n",
+    "\n",
+    "        return (return_value, rtn_params)\n",
+    "\n",
+    "        raise CIMError(CIM_ERR_METHOD_NOT_AVAILABLE)\n",
```

Comment:
Before merging this, I think we should flesh out which checks are done by the provider dispatcher in order to keep the provider simple. I have created issue #2326 for that.

Reply:
Done and documented for the method provider.   Note that I documented what validations we have in the pre-execution tests for the method provider in issue #2326. Note that we also don't test whether params are valid OUT/required in the response either.  The only one I can see that is missing would be test for valid param names and whether parameters are required/IN on input. To be honest we never did that in Pegasus but left it to the server method writer.  

Dropped the line with raise CIMError... from the code.   At this point the InvokeMethod itself is about 4 lines of code including displaying the input parameters.

Reply:
NOTE: I have not writted a user-defined instancewriteprovider for the notebook.